### PR TITLE
Factor-out userspace live patching macros

### PR DESCRIPTION
**Pull request by giulianobelinassi**
Userspace live patching requires that packages are built with special flags and artifacts to be packed for later usee. Instead of spawning code for this in every .spec file, a much better way is to factor out everything in a common file.

The way it is intended to be used is the following: packages that userspace livepatching support is desired should have defined in its begining:

%define livepatchable %{can_enable_livepatching}

Then it should append %{cflags_livepatching} on CFLAGS, and finally append %pack_ipa_dumps on %install.

Co-authored-by: @marxin

### CODE REVIEW

**Comment by marxin** [comment]
I'm thinking about a bit simplified usage of the macro, what about specifying the following in a spec file:

```
%define livepatchable 1
...
%pack_ipa_dumps
```

A conditional definition of `%{cflags_livepatching}` can be done even without Lua:
```shell
%cflags_livepatching %( \
  if [ %{livepatchable} ] ; then \
    if [ %{_arch} == "x86_64" ]; then \
      echo "-fpatchable-function-entry=16,14 -fdump-ipa-clones"; \
    else \
      echo ""; \
    fi \
  fi)
```

and then I would recommend, similarly to `%_lto_cflags`, add `%cflags_livepatching` unconditionally to `%optflags`:

https://build.opensuse.org/projects/openSUSE:Factory/prjconf:
`Optflags: * -O2 -Wall -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -Werror=return-type %%{?_lto_cflags} %%{?cflags_livepatching}
`

where the macro would be conditionally (based on `livepatchable` and arch) expanded to some options or to an empty string.
By doing that, you don't need to append `%{cflags_livepatching}` in each spec file.
What do you think?

File: suse_macros.in [hunk 1]
```diff
@@ -315,3 +315,28 @@ Provides translations for the \"%{-n:%{-n*}}%{!-n:%{name}}\" package.
...
+# Expand to the necessary livepatching compilation flags if the macro
+# livepatching is defined in the spec file.
+%cflags_livepatching %{lua: if rpm.expand("%livepatchable") == "1" then if rpm.expand("%_arch") == "x86_64" then print("-fpatchable-function-entry=16,14 -fdump-ipa-clones") end end}
+
+# Ipa-clones are files generated by gcc which logs changes made across
+# functions, and we need to know such changes to build livepatches
+# correctly. These files are intended to be used by the livepatch
+# developers and may be retrieved by using `osc getbinaries`.
+%pack_ipa_dumps \
+  %if %{with livepatching} || 0%{?livepatchable} \
```
Comment (marxin) [inline suse_macros.in:334]: Please remote here the `%{with livepatching}` leftover.

**Review by marxin — CHANGES_REQUESTED** [review]

File: suse_macros.in (hunk 1 above)
Comment (giulianobelinassi) [reply suse_macros.in]: @marxin done.

**Review by marxin — APPROVED** [review]

**Comment by giulianobelinassi** [comment]
ping 2

**Comment by darix** [comment]
why not justa have livepatching-macros package which you then can build require? that way we do not have to ship them in the default macros. and you can easily move it between distros even by just updating the livepatching-macros package in the other distros.

**Comment by giulianobelinassi** [comment]
Revoking this in favor of the `ulp-macros` package.

**Pull request closed without merging**